### ISSUE TITLE
Slow Datatable performance

### ISSUE THREAD

**Issue by AleksueiR** [fgpv-vpgf/fgpv-vpgf#2297]
​Datatable performance has dissolved into :poop: over the last few months and it's almost unusable even for smaller datasets (~1k records) in Chrome (it's unusable in IE, period). Large sets take a long time to build and display the table in addition to horrible scrolling performance.

The amount of data loaded into the table does matter, but Datatables has a infinite scroller example with 50k rows and it's pretty snappy: https://datatables.net/extensions/scroller/examples/initialisation/large_js_source.html

If you load `Plant Hardiness Zones` set, it's crawling:

![slow_table](https://user-images.githubusercontent.com/2285779/28924162-bec0d962-782e-11e7-9e5d-bab018bc8e98.gif)

#### Details and Zoom to buttons

These buttons are generated manually as Angular Material directives and inject them into the table using cells' `render` function. It's not possible to create just a single directive and reuse it for each row - a new one needs to be created for each row, rendered, stringified, and injected into the table as html to be re-rendered again as a button. This happens when the table is scrolled. A lot.

#### Truncating cell values

Truncating long values not fitting in cells happen in the `render` function at the DOM level with manually splitting strings and re-inserting them into the DOM (quite expensive). All string values are also run through the `linkify ` to make any present URLs clickable.

#### Table Keyboard Navigation

There is a custom functionality allowing keyboard navigation of the table (not that it's actually useful now since the table is so slow anyway). Observe:

![slow_table_keybaord](https://user-images.githubusercontent.com/2285779/28924174-c4e2eb28-782e-11e7-9084-83e508ab71fa.gif)

It seems, the code setting up keyboard support runs on every table draw (this happens a lot when you scroll) and it sets up new listeners on table elements. Listeners are piling up and slowing things down, also eating at memory.

#### Initial Table Rendering

`Plant Hardiness Zones` tables takes about 43 seconds to show up after all the data is loaded, and about 30 seconds more to become responsive. The first 43 seconds are spent in the `onTableInit` function on setting up filters and global search.

#### Knick-Knacks

Mostly inefficient jQuery selectors, not storing selection results and running the search again, using JS to hide DOM elements which would be more efficiently done with some CSS, that type of things. This is not a bottleneck, but it contributes to the overall sluggishness of the table.

#### Alignment of Quantum field

Insufficient blood muffins levels causes quantum fluctuations[.](https://xkcd.com/323/)

### What might help

- simplify process of building directives for `details` and `zoom to` buttons; 
  - use a single directive if possible and supply it with a number of row, or 
  - pre-cook them into the table data (avoids using `render` function, but adds more time on table initialization)
  - replace with simple, non-angular button with custom or no tooltips
- truncate cell value with CSS, use `title` attribute for tooltip, maybe make columns resizable
- refactor keyboard navigation to remove redundancies
- simplify filter/search setup if possible
- remove jQuery inefficient code
- eat more muffins

#### Last resort

If after all this, the table is still ~~:hankey:~~ :turtle:, go back to paginating results :cry::sob:.

Thoughts, anyone?

**Comment by milesap** [comment]
I vote for re-writing the entire table code base from scratch. Its getting hard to maintain, and we likely need to rip out a lot of stuff anyway to get it up to speed. It won't be easy given how much stuff has been piled in there already.

> Insufficient blood muffins levels causes quantum fluctuations.

😄 

I'm curious if the focus manager is also being a resource pig on large tables given it needs to manually find focusable elements.

**Comment by AleksueiR** [comment]
> I'm curious if the focus manager is also being a resource pig on large tables given it needs to manually find focusable elements.

I briefly considered checking that, but since it shouldn't affect scroll performance, decided not to :)

**Comment by milesap** [comment]
Before I forget - we're computing every cells width using canvas:

https://github.com/fgpv-vpgf/fgpv-vpgf/blob/master/src/app/ui/table/table-default.directive.js#L382

This is slow, but also not needed here since applying [the conditional CSS](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/master/src/app/ui/table/table-default.directive.js#L391) to all cells has the same effect.

**Comment by milesap** [comment]
I'm going to improve the DT performance as much as possible without completely re-writing the code base. We can then look at full optimization once the external API is ready by making it an extension (separate issue).

**Comment by milesap** [comment]
A few thoughts after spending some quality time with tables:

#### There are too many filters
We have 7 active filters on the table, which translates to 7 function calls per data item. So 100,000 records invokes 700,000 function calls. 

#### The filters are not optimized
A few filters are non-trivial - they either perform array searching, or create new objects like `Date`. 

#### Draw is called too often
There are 5 table draws before the table is even visible. So our 700,000 function calls just turned into 3,500,000.

#### Scroller extension
Its not clear why the filters are being called for every data item instead of as needed for rendering. Ideally the scroller extension would only invoke a filter `x` number of times where `x = display buffer * visible rows` which is currently `10 * 5 = 50`. This isn't the case.

#### Suggestions 
This **needs** to be redone from the ground up. We should look at:

1. Don't use filters at all, instead modify the table data directly when a filter is applied. This cuts loading time for the 100,000 row table from ~50 seconds to ~9 seconds while still using scroller.

2. Use one or two filters that are fast and use cached results, only re-computing when an actual filter is modified. Not as fast but a good step

3. After a certain number of rows (or table building time?), just disable filtering altogether (perhaps have a "turn on filtering anyways" button just in case they absolutely want to filter. This is likely a fine interim solution as well since it is trivia to implement).

Surprisingly when I removed the angular buttons and width calculations it not make a noticeable difference. I'll PR some of the changes just for demo purposes.

**Comment by milesap** [comment]
Commit: https://github.com/milesap/fgpv-vpgf/commit/5b310530ee4b63c0b306525e04453fa16c2acd81

Demo: http://fgpv.cloudapp.net/demo/users/milesap/dt-go-faster%21/prod/samples/index-samples.html

**Comment by jolevesq** [comment]
Even if I remove all the customization (filters, interactive buttons, columns customization and keyboard navigation) it is still slow. Sometimes I am even not able to open the table at all because it freezes.

I see there is a version 1.10.16 for Datatable. I will see if it helps.

**Comment by milesap** [comment]
DataTables should be fast by itself, though there may be things that really slow it down. They have a 50,000 record demo on their site that seems snappy (although its a basic table).

https://datatables.net/extensions/scroller/examples/initialisation/large_js_source.html

**Comment by jolevesq** [comment]
Yes it is but in their sample they don't have much stuff to do on draw (only 4 columns with no computation). Like you said previously the draw function append to often. We can increase scroller.displayBuffer value to reduce the number of draw. It is set to 10 right now. If I change it to 100, even with everything on, it is fast.

I agree there is too much draw (3 before table open) one because of ```self.table.scroller.measure(false);``` (with false option it can be remove) and 2 because of filters. Rewriting the whole thing is an option but for a quick fix the increase displayBuffer value can be use.

For this we can maybe calculate the best option with this in mind 
```
we tend to have fat tables which are hard to draw -> use small buffer https://datatables.net/reference/option/scroller.displayBuffer.
```

**Comment by milesap** [comment]
The odd thing in all of this is that if `scroller.displayBuffer` is 10 or 100 shouldn't matter too much. If DT has to draw 10 or 100 rows upfront shouldn't have a big time performance impact. Though what seems to be happening is that `scroller` is processing all rows - and this makes sense for filters since it needs to know which 10 or 100 rows it needs to draw. 

I'm fairly sure we need to reduce `draw` calls to only once up front, and combine all filters into one function. This "mega" filter would then need to be optimized (cache maybe) to quickly return filter results.

With `index-samples.html` opening the `crops` datatable, after the data has downloaded it takes me ~65 seconds to output the first set of rows with a `scroller.displayBuffer` of 100 :disappointed: 

The end plan is to remove the current DT implementation into an extention when the API is finished. We can have a lightweight DT extention that's fast and a more feature rich version like we do now (but hopefully optimized a bit more)

I'll give you more details about the DT removal from the core code base once I have more of the API completed and we can work on getting this and the lightweight DT into an extension.

**Comment by milesap** [comment]
To be revisited once enhanced table is rolled out.

**Comment by milesap** [comment]
New table in v3, datatable has been removed. No longer an issue, closing.
